### DISCUSSION TITLE
Is there a way to add or remove nodes in the source code?

### DISCUSSION THREAD

**Discussion by SKT1yang** [oxc-project/oxc#4970] [Q&A] [answered]
For example, adding an import statement or deleting a string literal expression. Here is my code implementation for modifying nodes:
<img width="974" alt="image" src="https://github.com/user-attachments/assets/2d264d5f-e8df-4948-bdc5-55f781d1693d">
<img width="1094" alt="image" src="https://github.com/user-attachments/assets/5130476c-ff4a-4a9e-8ce4-ae82cc8b949e">

**Comment by SKT1yang** [comment]
I searched for all the projects and code related to oxc that I could find on GitHub, but couldn't find any related implementations. My requirement is simple, which is to add an import statement in the JS code. Should I use oxc_transformer, oxc_traverse, or oxc_parser? Do you have a minimum example.

**Comment by DonIsaac** [comment]
cc: @overlookmotel

**Comment by overlookmotel** [comment, marked as answer]
Usually you'd use `oxc_ast::visit::VisitMut` for mutating AST (`oxc_traverse::Traverse` would also work, but that crate's API is likely to be reworked considerably in near future, and is only really intended for internal use within Oxc at present).

But for your simple use case, either of those is overkill. I'd suggest just parsing using `oxc_parser`, and then insert whatever nodes you want onto start of `program.body`. Please see the [parser example](https://github.com/oxc-project/oxc/blob/f8bb0222b3c9a23ceda0309cdfaabc40bbfd99f0/crates/oxc_parser/examples/parser.rs), and then you can create new nodes using `oxc_ast::AstBuilder`.

If you then need to print the AST back to JS text, see [codegen example](https://github.com/oxc-project/oxc/blob/f8bb0222b3c9a23ceda0309cdfaabc40bbfd99f0/crates/oxc_codegen/examples/codegen.rs).

Hope that helps.
